core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java

```diff
@@ -180,6 +179,21 @@ private void updateColumn(Types.NestedField field, Types.NestedField existingFie
     }
   }
 
+  private boolean ignorableUpdate(Type newType, Type existingType) {
+    if (newType.isPrimitiveType()) {
+      if (newType.equals(existingType)) {
+        return true;
+      }
+      if (existingType.typeId() == Type.TypeID.LONG && newType.typeId() == Type.TypeID.INTEGER) {
```

Comment:
Style nit:
```suggestion
      } else if (existingType.typeId() == Type.TypeID.LONG && newType.typeId() == Type.TypeID.INTEGER) {
```
Since there is no fall-through possible.

Reply:
Thanks for the suggestion, but I'm getting checkstyle rule violations when adding `else if`. This is what I tried, which appears to be logically correct
```java
    if (newType.isPrimitiveType()) {
      if (newType.equals(existingType)) {
        return false;
      } else if (existingType.typeId() == Type.TypeID.LONG
          && newType.typeId() == Type.TypeID.INTEGER) {
        return false;
      } else if (existingType.typeId() == Type.TypeID.DOUBLE
          && newType.typeId() == Type.TypeID.FLOAT) {
        return false;
      } else {
        return true;
      }
    }
    return false;
```

This is what I'm inclined to go with unless you have strong opinions on an alternative.
```java
    if (newType.isPrimitiveType()) {
      if (newType.equals(existingType)) {
        return false;
      }
      if (existingType.typeId() == Type.TypeID.LONG && newType.typeId() == Type.TypeID.INTEGER) {
        return false;
      }
      if (existingType.typeId() == Type.TypeID.DOUBLE && newType.typeId() == Type.TypeID.FLOAT) {
        return false;
      }
      return true;
    }
    return false;
```

Reply:
What is the checkstyle error?

Reply:
```shell
$ ./gradlew build -x test -x integrationTest
Configuration on demand is an incubating feature.

> Task :iceberg-open-api:validateRESTCatalogSpec
Validating spec /Users/rocco/Github/iceberg/open-api/rest-catalog-open-api.yaml
Spec is valid.

> Task :iceberg-aws:validateS3SignerSpec
Validating spec /Users/rocco/Github/iceberg/aws/src/main/resources/s3-signer-open-api.yaml
Spec is valid.

> Task :iceberg-core:checkstyleMain
[ant:checkstyle] [ERROR] /Users/rocco/Github/iceberg/core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java:189:14: Conditional logic can be removed. [SimplifyBooleanReturn]

> Task :iceberg-core:checkstyleMain FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':iceberg-core:checkstyleMain'.
> A failure occurred while executing org.gradle.api.plugins.quality.internal.CheckstyleAction
   > Checkstyle rule violations were found. See the report at: file:///Users/rocco/Github/iceberg/core/build/reports/checkstyle/main.html
     Checkstyle files with violations: 1
     Checkstyle violations by severity: [error:1]


* Try:
> Run with --scan to get full insights.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.10.2/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 12s
422 actionable tasks: 25 executed, 2 from cache, 395 up-to-date
```

Reply:
Shall we also consider DecimalType with different precision?

Reply:
i bet it wants you to smash that into the previous predicate. But i'm not sure that would be clearer for reading.

ie:
((double && float) || (long && int))

Reply:
@huaxingao 's Suggestion I think is also a good one to consider, but let's do that in a followup because I feel less confident about that one. I think it's fine though

Reply:
https://github.com/rocco408/iceberg/pull/1/files - Futzed around with this and I think this may be cleaner. I ended up renegating the function and used a switch statement. When I think about adding the decimal case, this seems like it will be cleaner

Reply:
Thank you @RussellSpitzer and @huaxingao, I pulled in those changes 🙇